src/NHibernate/Type/CollectionType.cs

```diff
@@ -359,36 +360,62 @@ public virtual object InstantiateResult(object original)
 			return result;
 		}
 
-		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache,
-											  ISessionImplementor session)
+		public virtual object ReplaceElements(object original, object target, object owner, IDictionary copyCache, ISessionImplementor session)
 		{
-			// TODO: does not work for EntityMode.DOM4J yet!
-			object result = target;
-			Clear(result);
+			var elemType = GetElementType(session.Factory);
+			var resultPc = target as IPersistentCollection;
```

Comment:
It would be nice to rename it `targetPc` for naming consistency, especially in the `ClearTargetsDirtyFlag` method.